### PR TITLE
#9164 - #9313 - Paths problems

### LINKED ISSUE #9164: 5ZoneTDV example file fatal due to incorrect path to schedule file in Ep-Launch V9.6
Issue overview
--------------
5ZoneTDV example file fatal out when run using EP-Launch Version 9.6 when trying to access Schedule:File object. EP-Launch is looking at incorrect path to the file. This file and others run fine in EnergyPlus V9.5. But it runs fine in V9.6 if you provide full path to the file in Schedule:File object.

### Details
Some additional details for this issue (if relevant):
 - EnergyPlus V9.6 (EP-Launch)

Error Message:
```
Severe  ** Schedule:File="ELECTDVFROMCZ06COM", File Name: "..\datasets\TDV\TDV_2008_kBtu_CTZ06.csv" not found.
   **   ~~~   **   Paths searched:
   **   ~~~   **     Current Working Directory: "C:\Data\EPlusPrograms\EnergyPlusV9-6-0\ExampleFiles\datasets\TDV"
   **   ~~~   **     EnergyPlus Executable Directory: "C:\Data\EPlusPrograms\datasets\TDV"
...
Fatal  ** ProcessScheduleInput: Preceding Errors cause termination.
```

EP-Launch should be looking at ```C:\Data\EPlusPrograms\EnergyPlusV9-6-0\datasets\TDV\``` instead of  ```C:\Data\EPlusPrograms\EnergyPlusV9-6-0\ExampleFiles\datasets\TDV```.

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (list location of defect file here)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### LINKED ISSUE #9313: FMU path problem for v9.6
Issue overview
--------------
In an earlier helpdesk ticket (16061) submitted by the NIST CONTAMW team, the author mentioned that there were some problems encountered in v9.5 and v9.6 in running an FMU. After some investigation, these two issues seem to be separated issues. The one with v9.6 will be is the first one to be encounter, which is also happending in the most up-to-date develop version (v22.1 IOFreeze & beyond). 

In v9.6 (and in v22.1 IOFreeze), it will first encounter is an FMI path problem with following message: 
```
   ** Severe ** ExternalInterface:FunctionalMockupUnitImport, FMU File Name: "ContamFMU.fmu" not found.
   ** ~~~ ** Paths searched:
   ** ~~~ ** Current Working Directory: "C:\Users\XXX\OneDrive - XXX\CONTAM\PRJ\coupled\mz320\to-Eplus-devTeam\run96\EPTEMP-00000001"
   ** ~~~ ** EnergyPlus Executable Directory: "C:\EnergyPlusV9-6-0"
   ** ~~~ ** "epin" Environment Variable: "C:\Users\wsdols\OneDrive - NIST\CONTAM\PRJ\coupled\mz320\to-Eplus-devTeam\run96\MZ320_V960"
   ** Fatal ** ExternalInterface/StopExternalInterfaceIfError: Error in ExternalInterface: Check EnergyPlus *.err file.
```

The problem is identified to be the path problem related to: 
https://github.com/NREL/EnergyPlus/blob/4fa778ec59f792656c87f75512e4ed21920f2859/src/EnergyPlus/ExternalInterface.cc#L880-L886
where the first parameter passed in has a forced cast to char* but the `WorkingFolder` variable is a path type and `WorkingFolder.c_str()` would yield wchar*. This will immediately result a cut of the string down to only the first character (with `\0 ` immediately following the first valid character). 

To fix this, `WorkingFolder.string().c_str()` needs to be used if a char* needs to be properly passed. The code should be revised to: 
```
                fmiEPlusInstantiateSlave((char *)state.dataExternalInterface->FMU(i).Instance(j).WorkingFolder.string().c_str(),
                                         &state.dataExternalInterface->FMU(i).Instance(j).LenWorkingFolder,
                                         &state.dataExternalInterface->FMU(i).TimeOut,
                                         &state.dataExternalInterface->FMU(i).Visible,
                                         &state.dataExternalInterface->FMU(i).Interactive,
                                         &state.dataExternalInterface->FMU(i).LoggingOn,
                                         &state.dataExternalInterface->FMU(i).Instance(j).Index);
```

After the fix the issue above, the v9.6 and v22.1-iofreeze versions are able to run the FMU if the FMU module is placed in one of the search folders. However, now they (v9.5, v9.6, and v22.1) now will all have a similar symptom related to FMU variables reading while running the FMU (see #9315). 

### Details
Some additional details for this issue (if relevant):
 - Platform (Operating system, version): all (or windows only?)
 - Version of EnergyPlus (if using an intermediate build, include SHA): v9.6 and v22.1-IOFreeze
 - Helpdesk ticket number 16061

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file added (list location of defect file [here](https://github.com/NREL/EnergyPlusDevSupport/tree/master/DefectFiles/9000s/9313))
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------
 - Fixes #9164 
      - For an explanation, please see https://github.com/NREL/EnergyPlus/issues/9164#issuecomment-1061663138
 - Fixes #9313
     - The description of commit https://github.com/NREL/EnergyPlus/commit/e5233cecfe8df506f32b0b568cd0ba3586f81cca is what matters most

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: src/EnergyPlus/DataSystemVariables.cc [hunk 1]
```diff
@@ -156,8 +152,9 @@ namespace DataSystemVariables {
             state.files.audit.ensure_open(state, "CheckForActualFilePath", state.files.outputControl.audit);
             std::string tmp;
 
+            // epin is passed from Epl-run.bat as the path to the IDF file minus its extension, so take the parent directory
             get_environment_variable(cInputPath1, tmp);
-            state.dataSysVars->envinputpath1 = fs::path(tmp);
+            state.dataSysVars->envinputpath1 = FileSystem::getParentDirectoryPath(fs::path(tmp));
```
Comment (jmarrec) [inline src/EnergyPlus/DataSystemVariables.cc:157]: Actual fix for #9164, fix is easy, finding the root cause was not (debugging something on windows that is launched via EPLaunch is a pain)

File: src/EnergyPlus/DataSystemVariables.cc [hunk 2]
```diff
@@ -175,14 +172,15 @@ namespace DataSystemVariables {
             {{InputFilePath, "Current Working Directory"},
              {state.dataStrGlobals->inputDirPath / InputFilePath, "IDF Directory"},
              {state.dataStrGlobals->exeDirectoryPath / InputFilePath, "EnergyPlus Executable Directory"},
-             {state.dataSysVars->envinputpath1 / InputFilePath, "\"epin\" Environment Variable"},
-             {state.dataSysVars->envinputpath2 / InputFilePath, "\"input_path\" Environment Variable"},
+             {state.dataSysVars->envinputpath1 / InputFilePath, R"("epin" Environment Variable)"},
```
Comment (jmarrec) [inline src/EnergyPlus/DataSystemVariables.cc:175]: cosmetic / modernize; use a raw string litteral

File: src/EnergyPlus/DataSystemVariables.cc [hunk 3]
```diff
@@ -175,14 +172,15 @@ namespace DataSystemVariables {
...
+             {state.dataSysVars->envinputpath2 / InputFilePath, R"("input_path" Environment Variable)"},
+             // These two only tested if `TestAllPaths` is true
              {state.dataStrGlobals->CurrentWorkingFolder / InputFilePath, "INI File Directory"},
-             {state.dataStrGlobals->ProgramPath / InputFilePath, "\"program\", \"dir\" from INI File"}}};
+             {state.dataStrGlobals->ProgramPath / InputFilePath, R"("program", "dir" from INI File)"}}};
 
-        std::size_t numPathsToNotTest = (state.dataSysVars->TestAllPaths) ? pathsToCheck.size() - 2 : pathsToCheck.size();
+        std::size_t numPathsToTest = (state.dataSysVars->TestAllPaths) ? pathsToCheck.size() : pathsToCheck.size() - 2;
```
Comment (jmarrec) [inline src/EnergyPlus/DataSystemVariables.cc:181]: Fix an unrelated logic error

File: src/EnergyPlus/DataSystemVariables.cc [hunk 4]
```diff
@@ -191,7 +189,7 @@ namespace DataSystemVariables {
                 std::pair<fs::path, std::string> currentPath(FileSystem::getParentDirectoryPath(FileSystem::getAbsolutePath(pathsToCheck[i].first)),
                                                              pathsToCheck[i].second);
                 bool found = false;
-                for (auto path : pathsChecked) {
+                for (auto &path : pathsChecked) {
```
Comment (jmarrec) [inline src/EnergyPlus/DataSystemVariables.cc:193]: Unrelated: avoid a copy on each iteration

File: src/EnergyPlus/ExternalInterface.cc [hunk 5]
```diff
@@ -871,13 +871,13 @@ void InstantiateInitializeFMUImport(EnergyPlusData &state)
...
 
     // Instantiate FMUs
-    for (i = 1; i <= state.dataExternalInterface->NumFMUObjects; ++i) {
-        for (j = 1; j <= state.dataExternalInterface->FMU(i).NumInstances; ++j) {
+    for (int i = 1; i <= state.dataExternalInterface->NumFMUObjects; ++i) {
+        for (int j = 1; j <= state.dataExternalInterface->FMU(i).NumInstances; ++j) {
+            auto folderStr = state.dataExternalInterface->FMU(i).Instance(j).WorkingFolder.string();
             state.dataExternalInterface->FMU(i).Instance(j).fmicomponent =
-                fmiEPlusInstantiateSlave((char *)state.dataExternalInterface->FMU(i).Instance(j).WorkingFolder.c_str(),
+                fmiEPlusInstantiateSlave((char *)folderStr.c_str(),
```
Comment (jmarrec) [inline src/EnergyPlus/ExternalInterface.cc:880]: Actual fix for #9313 . I used clang-query to find all calls to fs::path::c_str and made sure all others are fine.

File: src/EnergyPlus/ExternalInterface.cc [hunk 6]
```diff
@@ -871,13 +871,13 @@ void InstantiateInitializeFMUImport(EnergyPlusData &state)
     // Using/Aliasing
 
     // SUBROUTINE LOCAL VARIABLE DECLARATIONS:
-    int i, j; // Loop counters
 
     // Instantiate FMUs
-    for (i = 1; i <= state.dataExternalInterface->NumFMUObjects; ++i) {
-        for (j = 1; j <= state.dataExternalInterface->FMU(i).NumInstances; ++j) {
+    for (int i = 1; i <= state.dataExternalInterface->NumFMUObjects; ++i) {
+        for (int j = 1; j <= state.dataExternalInterface->FMU(i).NumInstances; ++j) {
```
Comment (jmarrec) [inline src/EnergyPlus/ExternalInterface.cc:877]: cosmetic/modernize

File: src/EnergyPlus/IOFiles.cc [hunk 7]
```diff
@@ -140,6 +140,7 @@ std::ostream::pos_type InputFile::position() const noexcept
 void InputFile::open(bool, bool)
 {
     file_size = fs::file_size(filePath);
+    // basic_fstream is a template, it has no problem with wchar_t (which filePath.c_str() returns on Windows)
```
Comment (jmarrec) [inline src/EnergyPlus/IOFiles.cc:143]: I added one comment here since that was one use of c_str() that wasn't obviously ok.

**Comment by jmarrec** [comment]
Running 5ZoneTDV.idf from ExamplesFiles folder in EPLaunch:

5ZoneTDV.audit, before and after fix

```diff
  Processing Schedule Input -- Start
not found (Current Working Directory)"=C:\EnergyPlusV22-1-0\ExampleFiles\datasets\TDV\TDV_2008_kBtu_CTZ06.csv
not found (IDF Directory)"=C:\EnergyPlusV22-1-0\ExampleFiles\datasets\TDV\TDV_2008_kBtu_CTZ06.csv
not found (EnergyPlus Executable Directory)"=C:\datasets\TDV\TDV_2008_kBtu_CTZ06.csv
-not found ("epin" Environment Variable)"=C:\EnergyPlusV22-1-0\ExampleFiles\datasets\TDV\TDV_2008_kBtu_CTZ06.csv
+found ("epin" Environment Variable)=C:\EnergyPlusV22-1-0\datasets\TDV\TDV_2008_kBtu_CTZ06.csv
-not found ("input_path" Environment Variable)"=C:\EnergyPlusV22-1-0\ExampleFiles\datasets\TDV\TDV_2008_kBtu_CTZ06.csv
```

File: src/EnergyPlus/DataSystemVariables.cc [hunk 8]
```diff
@@ -183,7 +183,8 @@ namespace DataSystemVariables {
         for (std::size_t i = 0; i < numPathsToTest; ++i) {
             if (FileSystem::fileExists(pathsToCheck[i].first)) {
                 foundFilePath = pathsToCheck[i].first;
-                print(state.files.audit, "{}={}\n", "found (" + pathsToCheck[i].second + ")", FileSystem::getAbsolutePath(foundFilePath).string());
+                print(state.files.audit, "found ({})={}\n", pathsToCheck[i].second, FileSystem::getAbsolutePath(foundFilePath).string());
```
Comment (jmarrec) [inline src/EnergyPlus/DataSystemVariables.cc:186]: Use print/fmt as it should. Deleted an extra trailing quote here too

```diff
-found ("epin" /path/to)"=...
+found ("epin" /path/to)=...
```

**Review by jcyuan2020 — COMMENTED** [review]
For issue 9313 fix: I've tested the defect files. Before the fix it gave errors about search paths; and now after the fix it is now working properly in identifying the fmu module and instance paths.

Also a light weight unit test may be possible to check the return value of fmiEPlusInstantiateSlave() call.

**Review by jcyuan2020 — COMMENTED** [review]
For the 9164 fix: I tested the defect testfile in a few different ways and the fix was able to find the intended search path is the folder is specified. 

One comment is about a special scenario where three search paths---"Current Work Directory", "EnergyPlus Executable Directory", and the "epin" Environment Variable---could yield more than one hit. 

One example is to specify path ".\DataSet\TDV\TDV_2008_kBtu_CTZ06.csv" and place the folder structure the idf working folder. Then "EnergyPlus Executable Directory" and "epin" Environment Variable will both be valid hits---and these two points to two different files that could be both valid (but could have different actual contents). 

So probably the path search order should be documented in the IORef that it will search the paths in a specific order and define which order should take precedence if there are multiple hits occurs---maybe it will search in an order and then stop once a hit is made. Not sure if such descriptions exist in the current documentation; if not, then might be worthwhile to add a few sentences about this.

**Comment by mjwitte** [comment]
@jmarrec Anything more to add here?

**Comment by mjwitte** [comment]
@jcyuan2020 There were some CI test failures here. A fix has been pushed. Please re-test for 9313 and 9164.

**Comment by jcyuan2020** [comment]
> 

On it.

**Comment by jcyuan2020** [comment]
The cases are retested ok.

**Pull request merged**